### PR TITLE
fix(ISV-6818): Update the SBOM merge process to prevent duplication.

### PR DESCRIPTION
**Pull request by haripate**
Fixes [ISV-6818](https://redhat.atlassian.net/browse/ISV-6818): when we merge Syft and Hermeto SBOMs (e.g. for Bootc images), the same package was sometimes listed twice because Syft and Hermeto format package IDs (purls) differently, so the merge treated them as different packages. The proposed changes make the merge use a single, normalized form of the package ID for non-Golang packages (dropping extra bits like qualifiers and subpath), so when both tools report the same package we keep only one entry. E2E and test data updates are in progress, and I’m still looking at other ways to handle merging.

[ISV-6818]: https://redhat.atlassian.net/browse/ISV-6818?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by Allda** [comment]
@haripate I am not sure if this PR adress the issue. I tried it on the sample SBOM that I extracted from real sbom with duplicated values and after executing your code it won't merge them.

Here is the data I used: https://gist.github.com/Allda/b99333b86714347b09d28701c7550519

There is a package `ttmkfdir` that is available in both sbom.

Please also include unit and integration tests that prove the duplicates are not produced on real examples.

**Comment by haripate** [comment]
> @haripate I am not sure if this PR adress the issue. I tried it on the sample SBOM that I extracted from real sbom with duplicated values and after executing your code it won't merge them.
> 
> Here is the data I used: https://gist.github.com/Allda/b99333b86714347b09d28701c7550519
> 
> There is a package `ttmkfdir` that is available in both sbom.
> 
> Please also include unit and integration tests that prove the duplicates are not produced on real examples.

Syft vs Hermeto dedupe compares a normalized PURL, and Syft’s RPM PURLs were carrying a `distro` qualifier while Hermeto’s for the same package typically were not (while other Hermeto-only qualifiers like checksum / repository_id were already ignored for the key), so the normalized strings diverged, the Syft row was never considered a duplicate of `Hermeto`, and both `ttmkfdir` entries stayed in the merge. The fix is to stop treating `distro` as part of that identity key so equivalent RPMs normalize to the same string, which lets `merge_by_prefering_hermeto` drop the Syft copy and keep Hermeto’s; the latest commit includes that change plus unit tests, and merge_sboms([syft.json], Hermeto.json) on this branch should now produce a single `ttmkfdir` for your example.

I also tested it locally with the data you provided earlier and here is the [output](https://gist.github.com/haripate/6a34753df66e655c7ce7de293397e281) I received from recent changes.

**Comment by Allda** [comment]
@haripate, before I do another round of review.. I see in the example merged output you shared the `ttmkfdir` coming from the Syft. Based on the Mobster logic, in case of duplicated packages, the Mobster prefers Hermeto. Is the output accurate?

**Comment by haripate** [comment]
> @haripate, before I do another round of review.. I see in the example merged output you shared the `ttmkfdir` coming from the Syft. Based on the Mobster logic, in case of duplicated packages, the Mobster prefers Hermeto. Is the output accurate?

I tried to test it on current Mobster merge workflow with the same test data but in that [outcome](https://gist.github.com/haripate/a41884f008adbc3510c6df17f6aa04cb), the `ttmkfdir` packiage data was coming from the Syft not Hermeto.

File: src/mobster/sbom/merge.py [hunk 1]
```diff
@@ -154,6 +154,30 @@ def wrap_as_spdx(items: list[dict[str, Any]]) -> list[SPDXPackage]:
...
+    identity_qualifiers = {"arch", "os", "classifier", "type", "epoch"}
+    meaningful_quals: dict[str, Any] = {}
+    for k, v in qualifiers.items():
+        if k not in identity_qualifiers:
+            continue
+        if k == "arch" and v == "noarch":
+            continue
+        meaningful_quals[k] = v
+
+    if purl_type == "golang" and meaningful_quals.get("type") == "module":
```
Comment (wcheang) [inline src/mobster/sbom/merge.py]: Why not also check this in the above loop before adding to the `meaningful_quals` dict, so all the checks are in one place?

File: src/mobster/sbom/merge.py [hunk 2]
```diff
@@ -328,41 +352,18 @@ def _is_hermeto_non_registry_dependency(component: SBOMItem) -> bool:
 def _unique_key_hermeto(component: SBOMItem) -> str:
     """
     Create a unique key from hermeto reported components.
-
-    This is done by taking a purl and removing any qualifiers and subpaths.
```
Comment (wcheang) [inline src/mobster/sbom/merge.py:332]: This should probably be updated instead of straight up removed.

File: src/mobster/sbom/merge.py [hunk 3]
```diff
@@ -328,41 +352,18 @@ def _is_hermeto_non_registry_dependency(component: SBOMItem) -> bool:
...
+
+    cleaned = _clean_identity_qualifiers(purl.qualifiers, purl.type)  # type:ignore[arg-type]
+    return purl._replace(qualifiers=cleaned, subpath=None).to_string()
 
 
 def _unique_key_syft(component: SBOMItem) -> str:
     """
     Create a unique key for Syft reported components.
-
-    This is done by taking a lowercase namespace/name, and URL encoding the version.
```
Comment (wcheang) [inline src/mobster/sbom/merge.py:352]: Same here with updating the comment. Also, some of it is still relevant, no?

Comment (haripate) [reply src/mobster/sbom/merge.py:352]: @wcheang Updated the comments and condition for qualifiers function as per your suggestion. Would you please review them now?

**Review by wcheang — APPROVED** [review]

**Review by jedinym — APPROVED** [review]
